File: CostServices/CostCalculateService.cs
using Entities;
using Interfaces;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace CostServices
{
    public class CostCalculateService : ICostCalculateService
    {
        private readonly IDiscountService discountService;
        private readonly ICostConfiguration costConfiguration;

        public CostCalculateService(
            IDiscountService discountService,
            ICostConfiguration costConfiguration
            )
        {
            this.discountService = discountService;
            this.costConfiguration = costConfiguration;
        }

        public decimal CalculatePayCheckCost(Employee emp)
        {
            var annualCost = this.CalculateAnnualCost(emp);
            return Math.Round(annualCost * 100 / 26) / 100;
        }

        private decimal CalculateAnnualCost(Employee emp)
        {
            var employeeCost = this.discountService.ApplyDiscount(emp.Name,
              this.costConfiguration.EmployeeAnnualCost);

            IEnumerable<string> dependents = emp.Children;
            if (!string.IsNullOrWhiteSpace(emp.SpouseName))
            {
                dependents = emp.Children.Concat(new[] { emp.SpouseName });
            }

            var dependentsCost = this.CalculateDependentsAnnualCost(dependents);
            return employeeCost + dependentsCost;
        }

        private decimal CalculateDependentsAnnualCost(IEnumerable<string> dependents)
        {
            return dependents.Select(d => this.CalculateDependentAnnualCost(d))
              .Aggregate(0m, (p, c) => p + c);
        }

        private decimal CalculateDependentAnnualCost(string name)
        {
            return this.discountService.ApplyDiscount(name, costConfiguration.DependentAnnualCost);
        }
    }
}

File: EmployeeCostCalculator/ClientApp/src/app/services/cost-calculate-service.ts
import { Injectable } from '@angular/core';
import { Employee } from '../models/Employee';
import { DiscountService } from './discount-service';
import { CostConfiguration } from './cost-configuration';
import { ICostCalculateService } from './ICostCalculateService';
import { of, Observable } from 'rxjs';


@Injectable({
  providedIn: 'any'
})
export class CostCalculateService implements ICostCalculateService {

  constructor(private discountService: DiscountService) {
  }

  calculatePayCheckCost(emp: Employee): Observable<number> {
    const annualCost = this.calculateAnnualCost(emp);
    return of(Math.round(annualCost * 100 / 26) / 100);
  }

  private calculateAnnualCost(emp: Employee): number {
    let employeeCost = this.discountService.applyDiscount(emp.name,
      CostConfiguration.employeeAnnualCost);

    let dependents = emp.children?.length ? [...emp.children] : [];
    const hasSpouse = !!emp.spouseName?.trim();
    if (hasSpouse) {
      dependents.push(emp.spouseName);
    }

    let dependentsCost = this.calculateDependentsAnnualCost(dependents);
    return employeeCost + dependentsCost;
  }

  private calculateDependentsAnnualCost(dependents: string[]): number {
    return dependents.map(d => this.calculateDependentAnnualCost(d))
      .reduce((p, c) => p + c, 0);
  }

  private calculateDependentAnnualCost(name: string): number {
    return this.discountService.applyDiscount(name, CostConfiguration.dependentAnnualCost);
  }
}
